src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -365,35 +415,92 @@ void TemporalMemory::learnOnSegments(
 
     if (isLearningSegment || isFromWinnerCell)
     {
-      adaptSegment(segment, activeSynapses, _connections);
+      adaptSegment(segment, activeSynapses, _connections,
+        permanenceIncrement_, permanenceDecrement_);
     }
 
     if (isLearningSegment)
     {
       Int n = maxNewSynapseCount_ - Int(activeSynapses.size());
 
-      set<Cell> cells = pickCellsToLearnOn(
-        n, segment, prevWinnerCells, _connections);
-      for (Cell presynapticCell : cells)
+      set<Cell> learningCells = pickCellsToLearnOn(
+        n, segment,
+        prevWinnerCells, _connections);
+
+      for (Cell presynapticCell : learningCells)
       {
-        _connections.createSynapse(segment, presynapticCell, initialPermanence_);
+        _connections.createSynapse(
+          segment, presynapticCell, initialPermanence_);
+      }
+    }
+  }
+
+  if (predictedSegmentDecrement_ > 0.0)
+  {
+    for (Segment segment : prevMatchingSegments)
+    {
+      bool isPredictedInactiveCell = (find(
+        predictedInactiveCells.begin(), predictedInactiveCells.end(),
+        segment.cell) != predictedInactiveCells.end());
+
+      vector<Synapse> activeSynapses(activeSynapsesForSegment(
+        segment, prevActiveCells, _connections));
+
+      if (isPredictedInactiveCell)
+      {
+        adaptSegment(segment, activeSynapses, _connections,
+          -predictedSegmentDecrement_, 0.0);
       }
     }
   }
 }
 
-tuple<vector<Segment>, set<Cell>> TemporalMemory::computePredictiveCells(
-  set<Cell>& _activeCells,
-  Connections& _connections)
+tuple<vector<Segment>, set<Cell>, vector<Segment>, set<Cell>>
+TemporalMemory::computePredictiveCells(
+  set<Cell>& _activeCells, Connections& _connections)
 {
+  map<Segment, int> numActiveConnectedSynapsesForSegment;
+  map<Segment, int> numActiveSynapsesForSegment;
   vector<Cell> activeCells(_activeCells.begin(), _activeCells.end());
+
   Activity activity = _connections.computeActivity(activeCells, connectedPermanence_, activationThreshold_);
 
+  Activity matchingActivity;
```

Comment:
Instead of computing it again here, can you just call `computeActivity` with `connectedPermanence = 0` and `activationThreshold = minThreshold`?